src/settings/project/mod.rs

```diff
@@ -13,43 +13,293 @@ use std::path::{Path, PathBuf};
+    pub workers_dot_dev: Option<bool>,
+    pub zone_id: Option<String>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Manifest {
+    pub account_id: String,
+    pub env: Option<HashMap<String, Environment>>,
     #[serde(rename = "kv-namespaces")]
     pub kv_namespaces: Option<Vec<KvNamespace>>,
+    pub name: String,
+    pub private: Option<bool>,
+    #[serde(rename = "type")]
+    pub project_type: ProjectType,
+    pub route: Option<String>,
+    pub routes: Option<HashMap<String, String>>,
+    pub webpack_config: Option<String>,
+    pub workers_dot_dev: Option<bool>,
+    pub zone_id: Option<String>,
 }
 
-impl Project {
+impl Manifest {
+    pub fn new(config_path: &Path) -> Result<Self, failure::Error> {
+        let config = read_config(config_path)?;
+
+        // check for pre 1.1.0 KV namespace format
+        let kv_namespaces: Result<Vec<config::Value>, config::ConfigError> =
+            config.get("kv-namespaces");
+
+        validate_kv_namespaces_config(kv_namespaces)?;
+
+        let manifest = config.try_into()?;
+        Ok(manifest)
+    }
+
+    pub fn get_target(
+        &self,
+        environment_name: Option<&str>,
+        release: bool,
+    ) -> Result<Target, failure::Error> {
+        if release && self.workers_dot_dev.is_some() {
+            failure::bail!(
+                "The --release flag is not compatible with use of the workers_dot_dev field"
+            )
+        }
+
+        if release {
+            message::warn("--release will be deprecated");
+        }
+
+        let environment = match environment_name {
+            Some(environment_name) => match &self.env {
+                Some(environment_table) => {
+                    let environment = environment_table.get(environment_name);
+                    match environment {
+                        Some(environment) => Some(environment),
+                        None => failure::bail!(format!(
+                            "{} Could not find environment with name {}",
+                            emoji::WARN,
+                            environment_name
+                        )),
+                    }
+                }
+                None => failure::bail!(format!(
+                    "{} There are no environments specified in your wrangler.toml",
+                    emoji::WARN
+                )),
+            },
+            None => None,
+        };
+
+        let deprecate_warning =
+            "Please specify the workers_dot_dev boolean in the top level of your wrangler.toml";
+        let wdd_failure = format!(
+            "{} Your environment should only include `workers_dot_dev` or `route`",
+            emoji::WARN
+        );
+
+        // TODO: deprecate --release, remove warnings and parsing
+        // switch wrangler publish behavior to act the same at top level
+        // and environments
+        // brace yourself, this is hairy
+        let workers_dot_dev = match environment {
```

Comment:
In general, I think this hairy section could be greatly simplified by using `if let` syntax instead of many nested matches: https://doc.rust-lang.org/rust-by-example/flow_control/if_let.html